Improved the explanation of the 5 equations in the Kane's Method docs.

moorepants opened the pull request:
- [ ] Add tests for new attributes.

@sympy/mechanics 

<!-- BEGIN RELEASE NOTES -->
* physics.mechanics
  * Improved `KanesMethod` documentation.
<!-- END RELEASE NOTES -->

moorepants committed:
Improved the explanation of the 5 equations in the Kane's Method docs.

jbm950 commented on doc/src/modules/physics/mechanics/kane.rst:
```diff
@@ -22,12 +22,37 @@ differentiated non-holonomic equations.
+constraints, that are linear in the generalized speeds. There are fewer
+equations than generalized speeds, and thus describe the relationship between
+the dependent and independent generalized speeds. Where there are :math:`m`
+dependent speeds and :math:`o-m` independent speed.
```
Though it can be inferred, `o` isn't actually defined until line 63 which comes after this section.

moorepants replied:
Yes, I'll define all of these above:

- o: number of generalized speeds
- m: number of non-holonomic constraints
- n:
- p:

The last two are related to number of indpendent speeds I think. We are using Kane's notation for that, I'll look in his book.

jbm950 replied:
"independent speed" -> "speed" should be "speeds"

jbm950 commented on doc/src/modules/physics/mechanics/kane.rst:
```diff
@@ -22,12 +22,37 @@ differentiated non-holonomic equations.
+The second equation describes the non-holonomic constraints, i.e. velocity
+constraints, that are linear in the generalized speeds. There are fewer
+equations than generalized speeds, and thus describe the relationship between
+the dependent and independent generalized speeds. Where there are :math:`m`
```
`m` is defined as dependent speeds here but on line 63 it is defined as motion constraints. Although the motion constraints are the cause of the dependent speeds, should only one of the two wordings be used for consistency? Or are `m` and `o` here completely different than those defined on line 63?

moorepants replied:
They are the same. The number of dependent speeds is equal to the number of constraints. For example, if you have 5 generalized speeds and 3 motion constraint equations that are linear in all 5 generalized speeds, then you can solve the 3 equations for three unknowns, thus 3 dependent speeds, leaving 2 independent speeds. I'll clarify.

jbm950 commented on doc/src/modules/physics/mechanics/kane.rst:
```diff
@@ -22,12 +22,37 @@ differentiated non-holonomic equations.
+be used to augment the independent dynamical equations if it is desired to
+solve for the dependent generalized speeds.
 
 Kane's method forms two expressions, :math:`F_r` and :math:`F_r^*`, whose sum
```
This part restates the equation you mention while defining equation number 4. Maybe change this part to say that equation number 4, the dynamical differential equation is what is rearranged to improve flow? Thoughts?

moorepants replied:
Will adjust.

jbm950 commented:
The equations are briefly labeled on lines 13-15. Some of these labels are repeated in your more detailed descriptions. Are the brief labels as necessary since more detailed descriptions are now added. If removed I'd suggest adding that the first equation represents the holonomic constraints. (this might be useful anyways because the part after your addition starts talking about holonomic constraints which isn't mentioned in your addition and could cause some confusion for someone still expanding their dynamics knowledge like me)

jbm950 commented:
Last thing, related to my above comment, the brief labels in lines 13-15 mention equation 3 and 4 as being a collected set equations whereas your added description refers to them in singular. I understand that it's due to the collection of equations being collected as vectors and matrices to be represented as a single equation. Thoughts?

Personally I'm for the added description with my only concern being whether or not the addition goes against the "this page is a brief description, if you want more details here's a book reference" feel that is established in the very top.

moorepants committed:
Major rewrite of the Kane prose documentation.

1. Added clarity to the equations and introduced more precise variable names.
2. Introduced Python code in first section to tie things together better.

moorepants commented:
Just did a major update to these docs. I need careful review of this as it is complicated and I'm sure I have errors.

moorepants committed:
Fixed errors.

moorepants commented:
>  I understand that it's due to the collection of equations being collected as vectors and matrices to be represented as a single equation. Thoughts?

Fixed.

> Personally I'm for the added description with my only concern being whether or not the addition goes against the "this page is a brief description, if you want more details here's a book reference" feel that is established in the very top. 

These docs need more explanation. You shouldn't have to look at Kane's book to understand how to use the `KanesMethod` class. You should only have to look at the book to understand the algorithm that is hidden in the classes methods. I've added a bunch to help improve things.

moorepants committed:
Fixed links [skip ci].

moorepants commented:
@angadhn Can you check my math here?

moorepants committed:
Changed M to o and fixed the matrix multiplication.

moorepants committed:
Updated the kane class to match the documentation, also switching to xreplace. [skip ci]

moorepants commented:
@jbm950 I've got some changes here. Started creating attributes that mirror some of the variable names with human understandable names among other things. Still needs work...

jbm950 commented on sympy/physics/mechanics/kane.py:
```diff
@@ -23,26 +25,34 @@ class KanesMethod(object):
+        Matrices of the generalized coordinates.
+    u, speeds : Matrix, shape(n, 1)
+        Matrices of the generalized speeds. For a non-holonomic system these
+        are ordered as [u_s, u_r].
```
What do s and r represent?

moorepants replied:
s: independent coordinates/speeds
r: dependent coordinates/speeds

This is the notation from Kane's book. I was think about using it in the math and internally to the class.

jbm950 commented on sympy/physics/mechanics/kane.py:
```diff
@@ -115,14 +124,50 @@ class KanesMethod(object):
+            The left hand side of the zero equal kinematical differential
+            equations. If None, the simple definition of u = q' is used.
+        q_dependent : ordered iterable, len(o)
+            The
```
I think this parameter definition is incomplete.

moorepants replied:
Yes, I'm still not done with the PR. Been pecking away at it when I have a few minutes.

jbm950 commented on sympy/physics/mechanics/kane.py:
```diff
@@ -115,14 +124,50 @@ class KanesMethod(object):
+            equations. If None, the simple definition of u = q' is used.
+        q_dependent : ordered iterable, len(o)
+            The
+        configuration_constraints : iterable of SymPy expressions, optional
```
I think this parameter is missing a definition.

jbm950 commented on sympy/physics/mechanics/kane.py:
```diff
@@ -818,16 +976,70 @@ def forcing_full(self):
     @property
+    def speeds(self):
+        """Returns the column matrix of generalized speeds."""
+        return self._q
```
I think this should return `self._u`

moorepants replied:
Yes, thanks. Fixed locally. Will push later.

jbm950 commented on sympy/physics/mechanics/kane.py:
```diff
@@ -132,62 +177,96 @@ def __init__(self, frame, q_ind, u_ind, kd_eqs=None, q_dependent=None,
         if not iterable(q_dep):
             raise TypeError('Dependent coordinates must be an iterable.')
+        if not iterable(q_ind):
+            raise TypeError('Generalized coordinates must be an iterable.')
```
Since q_dep and q_ind are both generalized coordinates should the strings read "Dependent generalized coordinates" and "Independent generalized coordinates"?

moorepants replied:
Yes, I think changing the language to include "generalized" is more precise.

jbm950 commented on sympy/physics/mechanics/kane.py:
```diff
@@ -132,62 +177,96 @@ def __init__(self, frame, q_ind, u_ind, kd_eqs=None, q_dependent=None,
+            raise TypeError('Generalized coordinates must be an iterable.')
         q_ind = Matrix(q_ind)
+
+        self._qind = q_ind
```
A quick search of the document says this is the only place `self._qind` appears. Is it being added as an attribute for later use?

moorepants replied:
This needs to be added as a property: `independent_coordinates`.

jbm950 commented on sympy/physics/mechanics/kane.py:
```diff
@@ -132,62 +177,96 @@ def __init__(self, frame, q_ind, u_ind, kd_eqs=None, q_dependent=None,
         if not iterable(u_dep):
             raise TypeError('Dependent speeds must be an iterable.')
+        if not iterable(u_ind):
+            raise TypeError('Generalized speeds must be an iterable.')
```
Same question as was raised on line 200 about independent/dependent generalized coordinates.

moorepants replied:
Will fix.

jbm950 commented on sympy/physics/mechanics/kane.py:
```diff
@@ -132,62 +177,96 @@ def __init__(self, frame, q_ind, u_ind, kd_eqs=None, q_dependent=None,
+        acc : iterable of SymPy expressions
+            The acceleration constraint expressions.
+
+        """
```
This docstring is really well done and I like the inclusion of the equation definitions!

moorepants replied:
Thanks. I'll probably stop at the kin diff equations and leave cleanup for fr/fstar and linearize for later. Or you can improve it.

jbm950 commented on sympy/physics/mechanics/kane.py:
```diff
@@ -784,9 +910,26 @@ def auxiliary_eqs(self):
-            raise ValueError('Need to compute Fr, Fr* first.')
+            msg = ('kanes_equations() must be run before the mass matrix is '
+                   'available.')
+            raise ValueError(msg)
```
```
if not self._fr or not self._frstar:
             raise ValueError('Need to compute Fr, Fr* first.')
```
This line also appears in mass_matrix_full, forcing and forcing full. I think it should receive the same treatment you gave `none_handler` where it is factored into a function call to reduce code redundancy. Would such a function be better as a method of `KanesMethod` or at the top of the module like `_none_handler`?

moorepants replied:
I can make these static methods of KanesMethod.

moorepants committed:
More cleanups and updates.

moorepants committed:
More properties.

jbm950 commented on sympy/physics/mechanics/kane.py:
```diff
@@ -15,48 +15,87 @@
 
 class KanesMethod(object):
-    """Kane's method object.
+    """This class is used automatically form the equations of motion of a
```
This class is used "to" automatically...

jbm950 commented on sympy/physics/mechanics/kane.py:
```diff
@@ -15,48 +15,87 @@
+    holonomic : boolean
+        True if the system is a holonomic system.
+    nonholonomic : boolean
+        True if the system is a nonholonomic system.
```
Isn't a system holonomic if it is not nonholonomic? Thus if they are both included there are two different ways to access the same information? I.e. `if holonomic:` is the same as `if not holonomic`

I see one of the advantages of python over other programming languages is it's attempt at a "one method of doing things" for simplicity sake. Otherwise there are too many variations to achieve the same task and at some point it becomes simpler to rewrite code instead of reusing/editing code (it's my understanding this is the state that the Perl language is in). For the same reason I think if we change from u -> speeds, q -> coordinates etc we should deprecate the old attributes. Long run it would reduce clutter and complexity.

jbm950 commented on sympy/physics/mechanics/kane.py:
```diff
@@ -15,48 +15,87 @@
+        and kinematic differential equations.
+    forcing_full : Matrix, shape(2n, 1)
+        The right hand side or "forcing vector" of the full first order form of
+        the implicit dynamic and kinematic differential equations.
```
I think `self.rhs()` should be mentioned somewhere in the docstring as the method of obtaining the equations of motion in explicit form. I don't know exactly where it would go though because it is not an attribute like the other but rather a method so that it's potentially long calculations don't bog down code unless it is required.

jbm950 commented on sympy/physics/mechanics/kane.py:
```diff
@@ -121,48 +160,60 @@ class KanesMethod(object):
-        u_dependent : ordered iterable, len(m)
-            The dependent speeds.
-        velocity_constraints : iterable of SymPy expressions, optional
+            equations. (TODO: check if following statement is true) If None,
```
The absence of kd_eqs is handled on line 435 (green side) where it simply initializes empty matrices and sets `self._qdot_u_map` to `None`. `self._qdot_u_map` is tested against `None` when determining if there are kinematic differential equations to return from the `kinddiffdict` (or something like that) but is also tested on lines 333 and 350 (also green side) and it is there that I believe you can determine the truth of the statement.

jbm950 commented on sympy/physics/mechanics/kane.py:
```diff
@@ -121,48 +160,60 @@ class KanesMethod(object):
-        velocity_constraints : iterable of SymPy expressions, optional
+            equations. (TODO: check if following statement is true) If None,
+            the simple definition of u = q' is used.
+        q_dependent : ordered iterable functions of time, len(o), optional
```
ordered iterable "of" functions of time. Just for consistency with q_ind and u_ind

jbm950 commented on sympy/physics/mechanics/kane.py:
```diff
@@ -121,48 +160,60 @@ class KanesMethod(object):
+            The o dependent generalized coordinates.
+        configuration_constraints : iterable of expressions, optional
+            The left hand side of the zero equal holonomic constraints.
+        u_dependent : ordered iterable, len(m), optional
```
Design question. The independent speeds and coordinates are `q_ind` and `u_ind` where the dependent speeds and coordinates do not use the abbreviations (`q_dependent` and `u_dependent`). Should one style be chosen over the other for consistency between the two? (Worth noting the code later uses `q_dep` and `u_dep` internally in `_initialize_vectors()`)

jbm950 commented on sympy/physics/mechanics/kane.py:
```diff
@@ -188,16 +239,16 @@ def __init__(self, frame, q_ind, u_ind, kd_eqs=None, q_dependent=None,
         if not iterable(q_ind):
-            raise TypeError('Generalized coordinates must be an iterable.')
+            msg = 'Independent generalized coordinates must be an iterable.'
+            raise TypeError(msg)
```
Was moving the message to a separate variable to keep the line < 80 characters? My Github viewer doesn't show 80 characters so I can't tell just looking at it.

jbm950 commented on sympy/physics/mechanics/kane.py:
```diff
@@ -244,11 +295,12 @@ def _initialize_constraint_matrices(self, config, vel, acc):
+        vel : iterable of expressions
             The velocity constraint expressions.
-        acc : iterable of SymPy expressions
+        acc : iterable of expressions
```
Out of curiosity why is "Sympy" being removed here (and elsewhere in similar situations)?

jbm950 commented on sympy/physics/mechanics/kane.py:
```diff
@@ -331,8 +383,8 @@ def _initialize_constraint_matrices(self, config, vel, acc):
-            B_s = self._k_nh[:, :self.num_ind_speeds]
-            B_r = self._k_nh[:, self.num_ind_speeds:]
+            B_s = self._k_nh[:, :self.num_independent_speeds]
+            B_r = self._k_nh[:, self.num_independent_speeds:]
```
This goes along with the question up above about choosing a standard between abbreviating and not abbreviating. The abbreviated version makes for shorter lines of code but I like the "at a glance" clarity that the non abbreviated version brings.

jbm950 commented on sympy/physics/mechanics/kane.py:
```diff
@@ -343,44 +395,40 @@ def _initialize_constraint_matrices(self, config, vel, acc):
-
             # Substitution dictionaries for setting variables to zero.
-            u_zero = dict((i, 0) for i in u)
+            u_zero = dict((i, 0) for i in self.u)
```
I noticed when looking through a few weeks ago that all of the internal code in `KaneMethod` would use the private attribute over the public attribute with the only exceptions being u and q (though this may no longer be the case with the large volume of changes propose). I was wondering if using the private attributes would be better practice for the internal code or if there's no real stylistic difference.

jbm950 commented on sympy/physics/mechanics/kane.py:
```diff
@@ -843,31 +891,40 @@ def kanes_equations(self, bodies, loads=None):
                     issue=10945, deprecated_since_version="1.1").warn()
 
+        # TODO : Why is it optional to pass in the kinematic differential
+        # equations? I'm not sure what the use case is without them.
```
As mentioned in the first docstring the kinematic differential equations can be left out and it would be assumed that u = q' (assuming the code actually treats this). If that is the case, however, I would think that the formation of the equations of motion should still work.

jbm950 commented on sympy/physics/mechanics/kane.py:
```diff
@@ -843,31 +891,40 @@ def kanes_equations(self, bodies, loads=None):
 
     def rhs(self, inv_method=None):
-        """ Returns the system's equations of motion in first order form.
+        """Returns the system's equations of motion in first order form.
```
Returns the system's equations of motion in "explicit" first order form. For consistency in the labeling of forms done in the main docstring.

jbm950 commented on sympy/physics/mechanics/kane.py:
```diff
@@ -923,36 +925,57 @@ def kanes_equations(self, bodies, loads=None):
+            If None LU decomposition is used to solve for x'. If `CH`, Cholesky
+            decomposition is used. Cholesky decomposition can almost always be
+            used, as long as the mass matrix is semi-positive definite. The
+            other methods are not recommended as they operate stricyly on the
```
stricyly -> strictly?

jbm950 commented on sympy/physics/mechanics/kane.py:
```diff
@@ -923,36 +925,57 @@ def kanes_equations(self, bodies, loads=None):
     def rhs(self, inv_method=None):
-        """Returns the system's equations of motion in first order form.
+        """Returns the right hand side of the explicit first order full
+        differential equations of motion where:
```
"full differential equations of motion ~~where~~:" I think this would flow better.

jbm950 commented on sympy/physics/mechanics/kane.py:
```diff
@@ -923,36 +925,57 @@ def kanes_equations(self, bodies, loads=None):
-            self._rhs = (self.mass_matrix_full.inv(inv_method,
-                         try_block_diag=True) * self.forcing_full)
+            udot = (self.mass_matrix.inv(method=inv_method,
+                                         try_block_diag=True) * self.forcing)
```
Ooh neat. I know you've mentioned that this method can be slow in the past, so getting rid if a lot of the calculations is great!

jbm950 commented:
Overall I'm -1 for having multiple attributes to access the same information (like self.o == self.num_dependent_coordinates). I believe having this will reduce code reusability and increase required code maintenance.

jbm950 commented on doc/src/modules/physics/mechanics/kane.rst:
```diff
@@ -1,59 +1,273 @@
+
+where the total equation is::
+
+   >>> Equality(kane.mass_matrix * kane.u, kane.forcing)
```
Isn't the equation M*__udot__= F not "u"?

jbm950 commented on sympy/physics/mechanics/kane.py:
```diff
@@ -112,17 +160,65 @@ class KanesMethod(object):
+                 configuration_constraints=None, u_dependent=None,
+                 velocity_constraints=None, acceleration_constraints=None,
+                 u_auxiliary=None):
+        """Initializes a KanesMethod object.
```
I like the idea of putting the parameters in the `__init__()` docstring but it appears this documentation is not picked up by Sphinx (unless I'm missing something). This can be verified by going to [Kane's online docstring documentation](http://docs.sympy.org/latest/modules/physics/mechanics/api/kane_lagrange.html?highlight=kanesmethod#sympy.physics.mechanics.kane.KanesMethod) and doing a quick search on "Please read the online documentation." which is `__init__`'s current docstring. Doing so will show that this docstring is nowhere to be found. I would suggest then that the parameters be moved back to the docstring for the class overall.

moorepants replied:
Note to self: http://stackoverflow.com/questions/5599254/how-to-use-sphinxs-autodoc-to-document-a-classs-init-self-method

jbm950 commented on sympy/physics/mechanics/kane.py:
```diff
@@ -15,49 +13,99 @@
+    n, num_coordinates, num_speeds : integer
+        The total number of generalized coordinates and speeds.
+    q, coordinates : Matrix, shape(n, 1)
+        The generalized coordinates. These are ordered as [q_s, q_r]^T.
```
Is n = number of generalized speeds + number of generalized coordinates? In general isn't num_coordinates = num_speeds not necessarily true? If it is the combination of the two would coordinates be a matrix of shape(n, 1)?

moorepants replied:
Yes you are correct. This needs some distinction. n should just be the number of generalized speeds I believe. I need to look back at kane's book to see how we handles this.

jbm950 replied:
```python
Notes
=====

n : number of independent generalized coordinates
m : number of dependent generalized speeds and nonholomic constraints
o : number of dependent generalized coordinates and holonomic constraints
p : number of independent generalized speeds
```
You have this written in the `__init__()` docstring.

moorepants replied:
Ok. That seems correct.

jbm950 commented on sympy/physics/mechanics/kane.py:
```diff
@@ -112,17 +160,65 @@ class KanesMethod(object):
+            The left hand side of the zero equal kinematical differential
+            equations. (TODO: check if following statement is true) If None,
+            the simple definition of u = q' is used.
+        q_dependent : ordered iterable functions of time, len(o), optional
```
ordered iterable "of" functions of time

moorepants committed:
More doc updates.

moorepants committed:
Reordering and minor edits.

moorepants committed:
A bit more cleanup on the doc page for Kane.

moorepants committed:
Fixed merge conflicts.

moorepants committed:
Minor updates.

moorepants committed:
Fixed merge conflicts.

czgdp1807 commented:
Is this needed?

moorepants commented:
Yep.
